core/src/main/java/org/mule/runtime/core/execution/interception/InterceptorMessageProcessor.java

```diff
@@ -0,0 +1,204 @@
+      ((FlowConstructAware) intercepted).setFlowConstruct(flowConstruct);
+    }
+    initialiseIfNeeded(intercepted, true, muleContext);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void start() throws MuleException {
+    startIfNeeded(intercepted);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void stop() throws MuleException {
+    stopIfNeeded(intercepted);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void dispose() {
+    disposeIfNeeded(intercepted, logger);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void setFlowConstruct(FlowConstruct flowConstruct) {
+    this.flowConstruct = flowConstruct;
+  }
+
+  /**
+   * @return the {@link ProcessingType} of the intercepted {@link Processor}.
+   */
+  @Override
+  public ProcessingType getProcessingType() {
+    return intercepted.getProcessingType();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Event process(Event event) throws MuleException {
+    final ComponentIdentifier componentIdentifier = (ComponentIdentifier) this.getAnnotations().get(ANNOTATION_NAME);
+    Map<String, Object> parameters = ((Map) this.getAnnotations().getOrDefault(ANNOTATION_PARAMETERS, emptyMap()));
+    Map<String, Object> resolvedParameters = new HashMap<>();
+    for (Map.Entry<String, Object> entry : parameters.entrySet()) {
+      Object value;
+      String paramValue = (String) entry.getValue();
+      if (muleContext.getExpressionManager().isExpression(paramValue)) {
+        value = muleContext.getExpressionManager().evaluate(paramValue, event, flowConstruct).getValue();
+      } else {
+        value = valueOf(paramValue);
+      }
+
+      resolvedParameters.put(entry.getKey(), value);
+    }
+
+    Optional<ProcessorInterceptorCallback> interceptorCallbackOptional = muleContext.getProcessorInterceptorManager()
+        .retrieveInterceptorCallback(componentIdentifier);
+      if(interceptorCallbackOptional.isPresent()) {
+        ProcessorInterceptorCallback processorInterceptorCallback = interceptorCallbackOptional.get();
+        try {
+          processorInterceptorCallback.before(event, resolvedParameters);
+          Event resultEvent;
+          if (processorInterceptorCallback.shouldInterceptExecution(event, resolvedParameters)) {
+            resultEvent = processorInterceptorCallback.getResult(event);
+          } else {
+            resultEvent = intercepted.process(event);
+          }
+          processorInterceptorCallback.after(resultEvent, resolvedParameters);
+          if (next != null) {
+            return next.process(resultEvent);
+          }
+          return resultEvent;
+        } catch (MuleException e) {
+          throw e;
+        } catch (Exception e) {
+          throw new MuleRuntimeException(e);
+        }
+      }
+      return intercepted.process(event);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Publisher<Event> apply(Publisher<Event> publisher) {
```

Comment:
This will cause all intercepting processors (and the rest of the chain) to block. I assume we don't want this? To avoid this, implement apply, and delegate and block from process.